extension/jaegerremotesampling/internal/remote_strategy_cache.go

```diff
@@ -0,0 +1,134 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/jaegerremotesampling/internal"
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/jaegertracing/jaeger/thrift-gen/sampling"
+	"github.com/tilinna/clock"
+)
+
+type serviceStrategyCache interface {
+	get(ctx context.Context, serviceName string) (*sampling.SamplingStrategyResponse, bool)
+	put(ctx context.Context, serviceName string, response *sampling.SamplingStrategyResponse)
+	Close() error
+}
+
+// serviceStrategyCacheEntry is a timestamped sampling strategy response
+type serviceStrategyCacheEntry struct {
+	retrievedAt      time.Time
+	strategyResponse *sampling.SamplingStrategyResponse
+}
+
+// serviceStrategyTTLCache is a naive in-memory TTL serviceStrategyTTLCache of service-specific sampling strategies
+// returned from the remote source. Each cached item has its own TTL used to determine whether it is valid for read
+// usage (based on the time of write).
+type serviceStrategyTTLCache struct {
+	itemTTL time.Duration
+
+	stopCh chan struct{}
+	rw     sync.RWMutex
+	items  map[string]serviceStrategyCacheEntry
+}
+
+// Initial size of cache's underlying map
+const initialRemoteResponseCacheSize = 32
+
+// Minimum time between runs of "clean up" of all cache-allocated data.
+// Used if the itemTTL is lower than this.
+const minCleanupSchedulingPeriod = 5 * time.Minute
+
+func newServiceStrategyCache(itemTTL time.Duration) serviceStrategyCache {
+	result := &serviceStrategyTTLCache{
+		itemTTL: itemTTL,
+		items:   make(map[string]serviceStrategyCacheEntry, initialRemoteResponseCacheSize),
+		stopCh:  make(chan struct{}),
+	}
+
+	// Launches a "cleaner" goroutine that naively blows away cached data
+	cleanerSchedulingPeriod := itemTTL
+	if itemTTL < minCleanupSchedulingPeriod {
```

Comment:
so here you're saying that we will not clean anything for a minimum of 5 minutes, no matter what TTL we have set.
Can we make this configurable instead?

Reply:
I guess I'm more taking issue with the comments line 92 to 96 that states that the cleanup interval is proportional to TTL, when it's not if TTL is less than 5 minutes.

It also means that there is a window of 5 minutes where the buffer can grow unbounded and no one can do much about it.

You start this cache with 32 entries - what's the expected type of volume you'd see go through that cache?

Reply:
The cache is keyed off of service names; in practice I would expect the number of items to be low (hundreds or thousands of values). This approach is defending against rarely queried items remaining in the cache and never getting expired out (replaced) if they are not queried. About volume of requests per collector, I expect it to be a function of the number of instrumented services, their respective agent configuration that controls agent polling, and the number of collectors deployed in the environment. I think I'll tweak it to just run on the configured TTL to simplify this in light of this conversation.

Reply:
How's what I just pushed?

It:
* Runs the cleaner at a frequency equal to the per item cache TTL
* Prunes stale items
* Recreates the map to prevent strictly increasing map growth (as unlikely as that may be given the expectations about service name cardinality in a typical deployment)

Hoping it's a bit more intuitive while still fulfilling my intent about managing garbage associated w/ the cache.

Reply:
> It also means that there is a window of 5 minutes where the buffer can grow unbounded and no one can do much about it.

BTW: Regarding this part of the concerns, the default behavior is no caching (it's inert/off by default) and somebody seeing problems created by caching can just turn it off via a extension config change.

And in case it was overemphasized, this thing growing unbounded would mean receiving a steady stream of cache-missing queries meaning you have an extremely large set of service _names_ (not just pods/host, but service identities) – a scenario I consider very unlikely but still defending against by this change.

Reply:
I'm gonna remove the comment that even mentions unbounded growth because it sounds a lot worse than it is.